apis/core/v1/gateway_types.go

```diff
@@ -0,0 +1,78 @@
+package v1
+
+import (
+	configv1 "github.com/rancher/opni/pkg/config/v1"
+	opnimeta "github.com/rancher/opni/pkg/util/meta"
+	"google.golang.org/protobuf/proto"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type GatewaySpec struct {
+	Replicas              *int32                      `json:"replicas,omitempty"`
+	Image                 *opnimeta.ImageSpec         `json:"image,omitempty"`
+	AgentImageTagOverride string                      `json:"agentImageTagOverride,omitempty"`
+	NatsRef               corev1.LocalObjectReference `json:"natsCluster"`
+	ServiceType           corev1.ServiceType          `json:"serviceType,omitempty"`
+	ServiceAnnotations    map[string]string           `json:"serviceAnnotations,omitempty"`
+	NodeSelector          map[string]string           `json:"nodeSelector,omitempty"`
+	Tolerations           []corev1.Toleration         `json:"tolerations,omitempty"`
+	Affinity              *corev1.Affinity            `json:"affinity,omitempty"`
+	ExtraVolumeMounts     []opnimeta.ExtraVolumeMount `json:"extraVolumeMounts,omitempty"`
+	ExtraEnvVars          []corev1.EnvVar             `json:"extraEnvVars,omitempty"`
+
+	//+kubebuilder:validation:Schemaless
+	//+kubebuilder:pruning:PreserveUnknownFields
+	//+kubebuilder:validation:Type=object
+	Config *configv1.GatewayConfigSpec `json:"config,omitempty"`
+}
+
+type GatewayStatus struct {
+	Image           string                      `json:"image,omitempty"`
+	ImagePullPolicy corev1.PullPolicy           `json:"imagePullPolicy,omitempty"`
+	ServiceName     string                      `json:"serviceName,omitempty"`
+	LoadBalancer    *corev1.LoadBalancerIngress `json:"loadBalancer,omitempty"`
+	Endpoints       []corev1.EndpointAddress    `json:"endpoints,omitempty"`
+	Ready           bool                        `json:"ready,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type Gateway struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              GatewaySpec   `json:"spec,omitempty"`
+	Status            GatewayStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
```

Comment:
If we're upgrading to v1 from v1beta1 should we include the storage version comment in one of the versions and also include conversion webhooks?